src/common/util/include/openvino/util/shared_object.hpp

```diff
@@ -9,36 +9,108 @@
 
 #pragma once
 
+#include <functional>
 #include <memory>
+#include <mutex>
+#include <thread>
 
 #include "openvino/util/util.hpp"
 
 namespace ov {
 namespace util {
+/**
+ * \brief Close shared object (library).
+ *
+ * Can be used as custom deleter in smart pointer.
+ */
+class SharedObjectCloser {
+public:
+    /**
+     * @brief Closes shared object at given pointer.
+     *
+     * @param shared_object pointer to shared object.
+     */
+    void operator()(void* shared_object) const;
+};
 
 /**
  * @brief Loads a library with the name specified.
- * @param path Full or relative path to the plugin library
+ *
+ * @param path             Full or relative path to the plugin library
+ * @param sh_object_closer custom closer of shared_object.
+ *
  * @return Reference to shared object
  */
-std::shared_ptr<void> load_shared_object(const char* path);
+std::shared_ptr<void> load_shared_object(const char* path,
+                                         std::function<void(void*)> sh_object_closer = SharedObjectCloser());
 
 #ifdef OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
 /**
  * @brief Loads a library with the wide char name specified.
- * @param path Full or relative path to the plugin library
+ *
+ * @param path             Full or relative path to the plugin library
+ * @param sh_object_closer Custom closer of shared_object.
+ *
  * @return Reference to shared object
  */
-std::shared_ptr<void> load_shared_object(const wchar_t* path);
+std::shared_ptr<void> load_shared_object(const wchar_t* path,
+                                         std::function<void(void*)> sh_object_closer = SharedObjectCloser());
 #endif  // OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
 /**
  * @brief Searches for a function symbol in the loaded module
+ *
  * @param shared_object shared object reference
- * @param symbolName Name of the function to find
+ * @param symbolName    Name of the function to find
+ *
  * @return A pointer to the function if found
  * @throws Exception if the function is not found
  */
 void* get_symbol(const std::shared_ptr<void>& shared_object, const char* symbolName);
 
+/**
+ * \brief Close shared object (library) with defer.
+ *
+ * Can be used as custom deleter in smart pointer.
+ */
+class SharedObjectDeferCloser {
+public:
```

Comment:
What benefit we can get if defer library close? 
It seems that SharedObjectDeferCloser  is not used in this PR, right?

Reply:
As PR is not finished, but using let say in more correct way the pimpl pattern this closer will be not required and is big chance that I will removed it later.

Reply:
Removed.